### PR TITLE
Add overloading resolution functions for fma

### PR DESCRIPTION
**Pull request by yxsamliu**

### CODE REVIEW

**Comment by b-sumner** [comment]
I'm just wondering what is wrong with reporting an error.  Can't we expect the developer to pass properly typed arguments?  Also, what will we do about half precision?

**Comment by yxsamliu** [comment]
std::numeric_limits<half>::is_specialized is false, therefore the overloading template has no effect on function calls with half argument.

It is a nuance for user to manually cast the arguments to double. This feature has been requested before.

cuda-clang has similar overloading resolution mechanism.

**Comment by b-sumner** [comment]
Are we sure we want the double function, even if all the arguments are bool?  Don't we want the float version if all the arguments can be losslessly cast to float?  Similarly, for fma(char, half, char) don't we want the half version?

**Comment by yxsamliu** [comment]
I updated the PR for better control of promoting to float fma or double fma. If float can hold all three arguments, the float fma is used, otherwise double fma is used.

For half type, since std::numeric_limits is not defined, the template does not apply. However, the standard overloading resolution still applies, so it does not make the situation worse.

**Comment by b-sumner** [comment]
Consuder these two cases:
    fma(float_var, (nt)3, float_var) --> fma(float, float, float)
    fma(float_var, (int)0x2000001, float_var) --> fma(double, double, double)  because 0x2000001 can't be exactly represented by a float

Now, a double can represent every [u]int, but not every [u]long, but with few enough non-zero bits, a float can exactly represent some [u]long.

I don't know how complex you want to make it, but the fma argument type should probably be chosen to exactly represent the original arguments as often as possible.

File: include/hip/hcc_detail/math_functions.h [hunk 1]
```diff
@@ -22,6 +22,7 @@ THE SOFTWARE.
 
 #pragma once
 
+#include "hip_fp16_math_fwd.h"
```
Comment (mangupta) [inline include/hip/hcc_detail/math_functions.h]: This causes build failures in the ```directed_tests/deviceLib/hipTestIncludeMath```. To fix this should either be included after ```hip/hcc_details/host_defines.h``` or we should modify ```hip_fp16_math_fwd.h``` to include ```host_defines.h```.

**Comment by yxsamliu** [comment]
updated PR. conflicts resolved.

**Pull request merged**